contrib/openssl-cmake/CMakeLists.txt

```diff
@@ -0,0 +1,1291 @@
+if(ENABLE_OPENSSL_DYNAMIC OR ENABLE_OPENSSL)
+    set(ENABLE_SSL 1 CACHE INTERNAL "")
+    set(OPENSSL_SOURCE_DIR ${ClickHouse_SOURCE_DIR}/contrib/openssl)
+    set(OPENSSL_BINARY_DIR ${ClickHouse_BINARY_DIR}/contrib/openssl)
+
+    set(OPENSSLDIR "/etc/ssl" CACHE PATH "Set the default openssl directory")
+    set(OPENSSL_ENGINESDIR "/usr/local/lib/engines-3" CACHE PATH "Set the default openssl directory for engines")
+    set(OPENSSL_MODULESDIR "/usr/local/lib/ossl-modules" CACHE PATH "Set the default openssl directory for modules")
+
+    add_definitions(-DOPENSSL_NO_SCTP -DOPENSSL_NO_KTLS -DOPENSSLDIR="${OPENSSLDIR}" -DENGINESDIR="${OPENSSL_ENGINESDIR}" -DMODULESDIR="${OPENSSL_MODULESDIR}" -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DDEVRANDOM="/dev/urandom")
+    target_compile_options(global-group INTERFACE "-Wno-deprecated-declarations")
+    target_compile_options(global-group INTERFACE "-Wno-poison-system-directories")
+
+    if(ARCH_AMD64)
```

Comment:
@bkuschel I know it has been a year or so but perhaps you remember something about the code here. I am trying to port ClickHouse from boringssl to OpenSSL ([PR](https://github.com/ClickHouse/ClickHouse/pull/56398)) and I am reusing most of the CMake code (which works for most of the platforms). What does not work (does not build) is aarch64 and riscv, and I think I need to find some variation of l. 15-26 and l. 101-119 for them. Since OpenSSL's native build is based on obscure Perl scripts with dozens of obscure [build flags](https://github.com/openssl/openssl/blob/master/INSTALL.md) and they reject any move to CMake: I wonder which invocation of OpenSSL's build was used as a base for this file or in other words / how were l. 15-26 / 101-119 determined? I am basically trying to reproduce what you did here but could unfortunately not find any reproduction steps. Any help is appreciated.

Reply:
@rschu1ze I need to rediscover this each time :) 
The original OpenSSL integration actually came from ClickHouse originally that subsequently removed. I took the original flag combination and reverse engineered the "configure" statement to make it work for each platform accordingly (and ran it as such). The original purpose was to build openssl into the clickhouse build in a way that was appropriate for development as it was intended that the openSSL build (.so) provided by the OS vendor would be used for production builds, so in that case we only needed the headers configured for clickhouse needs only. It may be possible to strip it down even more.
The most challenging part was finding the correct files to include for the compilation for a given platform build, as some algo and ciphers had platform-specific .s and .c/cpp files. This was a (painful) manual process.

I dug around in my notes and found these...  

_s390x_
```bash
../Configure \
    --prefix=/usr zlib enable-camellia enable-seed enable-rfc3779 enable-sctp enable-cms enable-md2 enable-rc5 enable-ktls enable-fips \
    no-mdc2 no-ec2m no-sm2 no-sm4 enable-buildtest-c++ shared linux64-s390x "-Wa,--noexecstack -Wa,--generate-missing-build-notes=yes -DPURIFY $RPM_LD_FLAGS" '-DDEVRANDOM="\"/dev/urandom\""'
```

_ppc64le_
```bash
../Configure --prefix=/usr enable-ec_nistp_64_gcc_128 zlib enable-camellia enable-seed enable-rfc3779 enable-sctp enable-cms enable-md2 enable-rc5 enable-ktls enable-fips no-mdc2 no-ec2m no-sm2 no-sm4 enable-buildtest-c++ shared linux-ppc64le "-Wa,--noexecstack -Wa,--generate-missing-build-notes=yes -DPURIFY $RPM_LD_FLAGS" '-DDEVRANDOM="\"/dev/urandom\""'
```
_x86_64_

```bash
../Configure --prefix=/usr enable-ec_nistp_64_gcc_128 no-zlib enable-camellia enable-seed enable-rfc3779 enable-sctp enable-cms enable-md2 enable-rc5 no-ktls enable-fips no-mdc2 no-ec2m no-sm2 no-sm4 enable-buildtest-c++ shared linux-x86_64 "-Wa,--noexecstack -Wa,--generate-missing-build-notes=yes -DPURIFY $RPM_LD_FLAGS" '-DDEVRANDOM="\"/dev/urandom\""'
```
_aarch65_
```
../Configure --prefix=/usr enable-ec_nistp_64_gcc_128 no-zlib enable-camellia enable-seed enable-rfc3779 enable-sctp enable-cms enable-md2 enable-rc5 no-ktls enable-fips no-mdc2 no-ec2m no-sm2 no-sm4 enable-buildtest-c++ shared linux-aarch64 "-Wa,--noexecstack -Wa,--generate-missing-build-notes=yes -DPURIFY $RPM_LD_FLAGS" '-DDEVRANDOM="\"/dev/urandom\""'